src/Shared/Contracts/INuGetProvider.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.Contracts;
+
+using Model;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.Versioning;
+using PackageManagers;
+using PackageUrl;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Version = System.Version;
+
+/// <summary>
+/// The interface to mock for NuGet.Protocol related methods we use.
+/// </summary>
+public interface INuGetProvider
+{
+    /// <summary>
+    /// Downloads a .nupkg file for a given <see cref="PackageURL"/>, and optionally extracts it.
+    /// </summary>
+    /// <param name="projectManager">The <see cref="NuGetProjectManager"/> this method is being called from, to use for file extraction.</param>
+    /// <param name="packageUrl">The <see cref="PackageURL"/> to download the .nupkg of.</param>
+    /// <param name="targetDirectory">The directory to save the contents to.</param>
+    /// <param name="doExtract">If the contents of the .nupkg should be extracted.</param>
+    /// <param name="cached">If the downloaded contents should be retrieved from the cache if they exist there.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to be used in the method call, defaults to <see cref="CancellationToken.None"/>.</param>
+    /// <returns>The location of the download, or null if the download failed.</returns>
+    Task<string?> DownloadNupkgAsync(NuGetProjectManager projectManager, PackageURL packageUrl, string targetDirectory, bool doExtract, bool cached = false, CancellationToken cancellationToken = default);
```

Comment:
In the new PR, I take in a `BaseProjectManager` instead. This is used because the `ExtractArchive` method is defined in the `BaseProjectManager` I would be ok with refactoring this somewhere else. Maybe not right now, but in the future. The only dependency `ExtractArchive` has in the class is the `TopLevelExtractionDirectory` which could easily be passed in as a parameter.